tests/functional/builtins/codegen/test_abi_decode_fuzz.py

```diff
@@ -0,0 +1,371 @@
+        bound = draw(st.integers(min_value=1, max_value=6))
+        return t(subtype, bound)
+    if t == DArrayT:
+        subtype = _go(skip=[TupleT])
+        bound = draw(st.integers(min_value=1, max_value=16))
+        return t(subtype, bound)
+
+    if t == TupleT:
+        # zero-length tuples are not allowed in vyper
+        n = draw(st.integers(min_value=1, max_value=6))
+        subtypes = [_go() for _ in range(n)]
+        return TupleT(subtypes)
+
+    if t in (BoolT, AddressT):
+        return t()
+
+    if t == IntegerT:
+        signed = draw(st.booleans())
+        bits = 8 * draw(st.integers(min_value=1, max_value=32))
+        return t(signed, bits)
+
+    if t == BytesM_T:
+        m = draw(st.integers(min_value=1, max_value=32))
+        return t(m)
+
+    raise RuntimeError("unreachable")
+
+
+@st.composite
+def data_for_type(draw, typ):
+    def _go(t):
+        return draw(data_for_type(t))
+
+    if isinstance(typ, TupleT):
+        return tuple(_go(item_t) for item_t in typ.member_types)
+
+    if isinstance(typ, SArrayT):
+        return [_go(typ.value_type) for _ in range(typ.length)]
+
+    if isinstance(typ, DArrayT):
+        n = draw(st.integers(min_value=0, max_value=typ.length))
+        return [_go(typ.value_type) for _ in range(n)]
+
+    if isinstance(typ, StringT):
+        # technically the ABI spec doesn't say string has to be valid utf-8,
+        # but eth-stdlib won't encode invalid utf-8
+        return draw(st.text(max_size=typ.length))
+
+    if isinstance(typ, BytesT):
+        return draw(st.binary(max_size=typ.length))
+
+    if isinstance(typ, IntegerT):
+        lo, hi = typ.ast_bounds
+        return draw(st.integers(min_value=lo, max_value=hi))
+
+    if isinstance(typ, BytesM_T):
+        return draw(st.binary(min_size=typ.length, max_size=typ.length))
+
+    if isinstance(typ, BoolT):
+        return draw(st.booleans())
+
+    if isinstance(typ, AddressT):
+        ret = draw(st.binary(min_size=20, max_size=20))
+        return "0x" + ret.hex()
+
+    raise RuntimeError("unreachable")
+
+
+def _sort2(x, y):
+    if x > y:
+        return y, x
+    return x, y
+
+
+@st.composite
+def _mutate(draw, payload, max_mutations=MAX_MUTATIONS):
+    # do point+bulk mutations,
+    # add/edit/delete/splice/flip up to max_mutations.
+    if len(payload) == 0:
+        return
+
+    ret = bytearray(payload)
+
+    # for add/edit, the new byte is any character, but we bias it towards
+    # bytes already in the payload.
+    st_any_byte = st.integers(min_value=0, max_value=255)
+    payload_nonzeroes = list(x for x in payload if x != 0)
+    if len(payload_nonzeroes) > 0:
+        st_existing_byte = st.sampled_from(payload)
+        st_byte = st.one_of(st_existing_byte, st_any_byte)
+    else:
+        st_byte = st_any_byte
+
+    # add, edit, delete, word, splice, flip
+    possible_actions = "adwww"
+    actions = draw(st.lists(st.sampled_from(possible_actions), max_size=MAX_MUTATIONS))
```

Comment:
this can also generate an empty list, no? Don't we want `min_size=1`?

Reply:
We want to test valid payloads too, we don't have to mutate the payload

Reply:
ok, i assumed we'd test valid payloads separately. i think it could be better to test happy paths separately. we could enmurate all type combinations up to certain depth